### PR TITLE
feat: add full supervisor policy runtime for tool approval and output formatter

### LINKED ISSUE #100: [Feature] Supervisor policy system v1 — intent guard, output formatter, playbook, routing, sequencing, approval + plan approval step
## What you want and why

The CUGA supervisor currently has no policy layer of its own. As multi-agent workflows grow in complexity, the supervisor needs the same governance primitives that exist for individual agents — plus orchestration-specific controls that govern how sub-agents are selected, ordered, and approved.

This feature tracks **v1 of the Supervisor Policy System**, implemented entirely through the Python SDK with full test coverage per policy type.

### Policy types in scope

| Policy | Description |
|--------|-------------|
| **Intent Guard** | Block or allow supervisor-level tasks before any sub-agent is dispatched |
| **Output Formatter** | Transform or gate the final response returned by the supervisor to the caller |
| **Playbook** | Define a structured multi-step workflow the supervisor must follow |
| **Agent Routing Policy** | Control which sub-agent(s) are eligible to handle a given task |
| **Agent Sequencing Policy** | Enforce ordering constraints on sub-agent execution (e.g. agent A must run before agent B) |
| **Agent Approval Policy** | Require human or automated approval before dispatching a specific sub-agent |

### Plan approval (separate, non-policy step)

In addition to the policy types above, add a dedicated **plan approval step** to the supervisor loop. This is not a policy — it is a first-class supervisor lifecycle hook that surfaces the sub-agent execution plan to an operator for review/approval before any sub-agent runs. This gives visibility into what the supervisor intends to do before it does it.

### SDK surface (high-level strawman)

```python
from cuga.supervisor import CugaSupervisor
from cuga.supervisor.policies import (
    SupervisorIntentGuard,
    SupervisorOutputFormatter,
    SupervisorPlaybook,
    AgentRoutingPolicy,
    AgentSequencingPolicy,
    AgentApprovalPolicy,
)

supervisor = CugaSupervisor(
    agents=[agent_a, agent_b, agent_c],
    policies=[
        SupervisorIntentGuard(blocked_patterns=["delete all", "drop table"]),
        SupervisorOutputFormatter(template="..."),
        SupervisorPlaybook(steps=["verify", "execute", "summarize"]),
        AgentRoutingPolicy(rules={"billing": agent_a, "hr": agent_b}),
        AgentSequencingPolicy(order=[agent_a, agent_b]),
        AgentApprovalPolicy(agents=[agent_c], requires_approval=True),
    ],
    plan_approval=True,  # surfaces sub-agent plan before execution
)
```

### Plan approval hook

When `plan_approval=True`, the supervisor pauses after building its execution plan and before dispatching any sub-agent. The plan (list of sub-agents + their inputs) is surfaced via the existing HITL mechanism for operator review. Execution resumes only after approval.

### Implementation scope

- New `cuga.supervisor.policies` module with a base class and one concrete class per policy type
- Supervisor loop updated to evaluate policies at the correct lifecycle points (pre-dispatch, post-output, routing, sequencing, approval)
- Plan approval as a separate supervisor lifecycle step (not coupled to the policy module)
- SDK-only for v1 — maybe just show enactment phases like what happened next to final answer as we did in existing system for policies

### Testing

Each policy type gets its own test file/class. Tests exercise the policy exclusively through the SDK (no UI, no integration harness). Coverage should include:

- Happy-path: policy allows execution to proceed
- Block/redirect path: policy halts or reroutes execution
- Edge cases per type (e.g. sequencing violation, routing miss, approval denied)
- Plan approval: plan is surfaced and execution is gated correctly

## Links or extra context

- Existing agent-level policy docs: https://docs.cuga.dev/docs/sdk/policies/
- Epic tracking the policy document parser (upstream): #99
- v1 is SDK-only; a future issue will add UI authoring and document-based policy loading

### PR DESCRIPTION
**Pull request by sami-marreed**
## Feature

Part of #100 · Epic #75

Relevant docs PR: https://github.com/cuga-project/docs/pull/3/changes

### Goal

Bring **full policy runtime support** to the standalone `CugaSupervisor` SDK so it matches `CugaAgent`: policies are not just stored via CRUD, they are **enforced at invoke time** — blocking bad intents, injecting playbooks/tool guides, pausing for human tool approval, and formatting output.

Before this PR, supervisor policies were only partially wired (intent guard / playbook / tool guide worked; tool approval could prompt but not resume; output formatter was storage-only). This closes that gap and adds e2e coverage for every policy type on the supervisor graph.

### Summary

- Wrap `CugaSupervisor` graph with HITL nodes (`SupervisorSDKCallback`, `SuggestHumanActions`, `WaitForResponse`) so tool approval pause/resume works via `invoke(..., action_response=...)`
- Apply output formatter policies at runtime in the supervisor callback (using `supervisor_metadata`)
- Fix policy context extraction for supervisor state (`input`, `supervisor_chat_messages`) and persist `supervisor_metadata` from prepare step
- Prefer existing `state.script` on tool approval resume
- Add invoke-level e2e tests for all supervisor policy types (intent guard, playbook, tool guide, tool approval, output formatter)

### Testing
- [x] Tested locally; tests pass
- [x] `uv run pytest src/cuga/sdk_core/tests/test_supervisor_policies.py -v` (17 passed, 3 consecutive runs)

### Docs
- Companion docs PR: https://github.com/cuga-project/docs/pull/3 (supervisor policies quick start + updated support matrix)

## Summary by CodeRabbit

* **Bug Fixes**
  * Tool approval execution now prefers existing script content when available.
  * Conversation context now uses additional fallbacks for user input and chat history.

* **New Features**
  * Output formatting supports configurable metadata storage and writes formatted answers plus formatter status to state.
  * Human-in-the-loop tool approval routing now returns to the correct UI surface and improves approval/resume behavior.

* **Tests**
  * Expanded supervisor policy end-to-end coverage, including approval confirm/deny and output-formatter transformation cases.

### CODE REVIEW

File: src/cuga/backend/cuga_graph/policy/output_formatter_utils.py [hunk 1]
```diff
@@ -17,7 +17,10 @@
 
 
 async def apply_output_formatter_policies(
-    state: AgentState, config: Optional[RunnableConfig] = None, context: str = "unknown"
+    state: AgentState,
+    config: Optional[RunnableConfig] = None,
+    context: str = "unknown",
+    metadata_key: str = "cuga_lite_metadata",
```
Comment (offerakrabi) [inline src/cuga/backend/cuga_graph/policy/output_formatter_utils.py:23]: is the `cuga_lite_metadata` a new default name? if so, maybe we should refrain from using `cuga_lite` as we wish to phase out this term

File: src/cuga/backend/cuga_graph/nodes/cuga_agent_core/policy/tool_approval_handler.py
Comment (offerakrabi) [inline src/cuga/backend/cuga_graph/nodes/cuga_agent_core/policy/tool_approval_handler.py:184]: This call now reuses [create_tool_approval_action()](vscode-webview://04lie157kj6r918rgg6vqe0bmkokfqcbp3kqhtg4rlu2hfmevap3/src/cuga/backend/cuga_graph/nodes/human_in_the_loop/followup_model.py:118) for supervisor tool-approval flows, but that helper currently creates an action whose return_to metadata is hard-coded to CugaLite. Since this path can now be reached from CugaSupervisor, UI/server clients that route or label approval responses based on the HITL action metadata may treat supervisor tool approvals as CugaLite approvals.

Please pass the correct return target from this call site, or make [create_tool_approval_action()](vscode-webview://04lie157kj6r918rgg6vqe0bmkokfqcbp3kqhtg4rlu2hfmevap3/src/cuga/backend/cuga_graph/nodes/human_in_the_loop/followup_model.py:118) accept a configurable return target with a default that preserves existing CugaLite behavior. For supervisor flows, this should produce return_to="CugaSupervisor", consistent with supervisor agent approvals.

see (followup_model.py at line 159) this is where `return_to` is hard-coded to always be `CugaLite`

File: src/cuga/sdk_core/tests/test_supervisor_policies.py [hunk 2]
```diff
@@ -235,3 +295,172 @@ async def fake_ensure():
...
+
+        result = await supervisor.invoke("Get Alice's user id")
+        metadata = (_graph_values(supervisor, result.thread_id).get("supervisor_metadata")) or {}
+        guides = metadata.get("guides") or []
+
+        assert guides
+        assert any(g.get("policy_name") == "Delegation Guide" for g in guides)
+
+    @pytest.mark.asyncio
+    async def test_e2e_tool_approval_requests_approval(self):
```
Comment (offerakrabi) [inline src/cuga/sdk_core/tests/test_supervisor_policies.py:400]: This test verifies the SDK resume happy path, but it does not validate the approval action payload that the UI/server consumes after the interrupt. Since the test manually constructs the approval response at [src/cuga/sdk_core/tests/test_supervisor_policies.py:439](vscode-webview://04lie157kj6r918rgg6vqe0bmkokfqcbp3kqhtg4rlu2hfmevap3/src/cuga/sdk_core/tests/test_supervisor_policies.py:439), it would not catch a malformed generated HITL action, such as the supervisor approval action using the wrong return target.

Please assert the stored HITL action from the interrupted state here, including the return target, action id, and relevant tool/code payload. It would also be useful to add a denial-path test to verify that a negative confirmation does not continue execution or delegation.

main suggestion here is not to actually load the ui, but you can still test the ui playloads

**Review by offerakrabi — COMMENTED** [review]
Review summary: the PR’s overall direction looks sound: it adds supervisor policy/HITL support in a way that fits the existing graph flow, improves supervisor policy context, and includes useful invoke-level tests. I did not find Critical issues. The main concern is around the supervisor tool-approval contract exposed to external clients: the runtime path works at the SDK level, but the generated approval action metadata needs to be validated and corrected for supervisor flows before merge.

Issues:
1. `cuga_lite_metadata` used as default key [`src/cuga/backend/cuga_graph/policy/output_formatter_utils.py:23`](src/cuga/backend/cuga_graph/policy/output_formatter_utils.py:23)
2. Important — Supervisor tool approval reuses an action helper that can emit the wrong return target; comment on [`src/cuga/backend/cuga_graph/nodes/cuga_agent_core/policy/tool_approval_handler.py:184`](src/cuga/backend/cuga_graph/nodes/cuga_agent_core/policy/tool_approval_handler.py:184).
3. Important — Supervisor approval tests cover SDK resume but not the UI/server-facing HITL action metadata or denial path; comment on [`src/cuga/sdk_core/tests/test_supervisor_policies.py:436`](src/cuga/sdk_core/tests/test_supervisor_policies.py:436).

File: src/cuga/sdk.py [hunk 3]
```diff
@@ -2861,6 +2861,103 @@ async def initialize(self):
...
+                        return Command(update=state.model_dump(), goto="SupervisorSubgraph")
+                    policy_name = (state.supervisor_metadata or {}).get("policy_name", "Tool Approval")
+                    state.final_answer = (
+                        f"❌ **Execution Cancelled**\n\nYou denied execution required by "
+                        f"**{policy_name}**. The supervisor will not proceed with this task."
+                    )
+                    state.execution_complete = True
+                    state.hitl_response = None
+                    state.sender = callback_name
+                    return Command(update=state.model_dump(), goto=END)
```
[the remark this answers is not in the document]
Comment (sami-marreed) [reply src/cuga/sdk.py:2917]: Fixed in 7b2c5b27 — denial path now sets hitl_action = None alongside hitl_response.

File: src/cuga/backend/cuga_graph/policy/output_formatter_utils.py (hunk 1 above)
Comment (sami-marreed) [reply src/cuga/backend/cuga_graph/policy/output_formatter_utils.py:23]: No rename in this PR — default matches existing AgentState.cuga_lite_metadata for CugaLite callers. Supervisor already passes metadata_key="supervisor_metadata". Docstring updated.

**Comment by sami-marreed** [comment]
Updates addressing review (HEAD `7b2c5b27`):

**Addressed**

- **#c4** — Fixed — `create_tool_approval_action` accepts `return_to`; supervisor passes `adapter.sender_name` (`CugaSupervisor`) (`followup_model.py`, `tool_approval_handler.py`)
- **#c5** — Fixed — E2E asserts generated `hitl_action` payload (`return_to`, `action_id`, tool/code fields); added denial-path test (`test_supervisor_policies.py`)
- **#c1** — Fixed — denial path clears `hitl_action` alongside `hitl_response` (`sdk.py`)
- **#c2** — Fixed — resume requires explicit `thread_id` before auto-generate (`sdk.py`, matches `CugaAgent.invoke`)

**Clarifications**

- **#c3** — No rename in this PR — default `metadata_key` matches existing `AgentState.cuga_lite_metadata` field; supervisor already passes `supervisor_metadata`. Docstring updated; full term phase-out is a separate refactor.

Ready for another look.

File: src/cuga/backend/cuga_graph/nodes/cuga_agent_core/policy/tool_approval_handler.py
Comment (sami-marreed) [reply src/cuga/backend/cuga_graph/nodes/cuga_agent_core/policy/tool_approval_handler.py:184]: Fixed in 7b2c5b27 — return_to now comes from adapter.sender_name (CugaSupervisor for supervisor flows).

File: src/cuga/sdk_core/tests/test_supervisor_policies.py (hunk 2 above)
Comment (sami-marreed) [reply src/cuga/sdk_core/tests/test_supervisor_policies.py:400]: Fixed in 7b2c5b27 — approval E2E now asserts hitl_action payload (return_to, action_id, tool fields). Added test_e2e_tool_approval_denial_cancels_execution.

**Review by offerakrabi — APPROVED** [review]
